File: 조성원/[LV2] N-Queen.js
function isAvailable(cols, row, col) {
  for (let i = 0; i < row; i++)
    if (cols[i] === col || Math.abs(cols[i] - col) === row - i) return false;

  return true;
}

function placeQueens(n, row, cols) {
  if (row === n) return 1;

  let count = 0;
  for (let col = 0; col < n; col++) {
    if (!isAvailable(cols, row, col)) continue;
    cols[row] = col;
    count += placeQueens(n, row + 1, cols);
    cols[row] = -1;
  }
  return count;
}

function solution(n) {
  const cols = Array(n).fill(-1);
  return placeQueens(n, 0, cols);
}

File: 김현경/[LV2] N-Queen.py
def queen(x, n, col, diag1, diag2):
    if x == n:
        return 1
    count = 0
    for y in range(n):
        if not col[y] and not diag1[x - y] and not diag2[x + y]:
            col[y] = True
            diag1[x - y] = True
            diag2[x + y] = True
            count += queen(x + 1, n, col, diag1, diag2)
            col[y] = False
            diag1[x - y] = False
            diag2[x + y] = False
    return count

def solution(n):
    col = [False] * n
    diag1 = [False] * (2 * n - 1)
    diag2 = [False] * (2 * n - 1)
    
    return queen(0, n, col, diag1, diag2)